python/lsst/daf/butler/core/uri/_butlerUri.py

```diff
@@ -0,0 +1,668 @@
+
+    All non-schemeless URIs have to internally use quoted paths. Therefore
+    if a new file name is given (e.g. to updateFile or join) a decision must
+    be made whether to quote it to be consistent.
+    """
+
+    # This is not an ABC with abstract methods because the __new__ being
+    # a factory confuses mypy such that it assumes that every constructor
+    # returns a ButlerURI and then determines that all the abstract methods
+    # are still abstract. If they are not marked abstract but just raise
+    # mypy is fine with it.
+
+    # mypy is confused without this
+    _uri: urllib.parse.ParseResult
+
+    def __new__(cls, uri: Union[str, urllib.parse.ParseResult, ButlerURI],
+                root: Optional[Union[str, ButlerURI]] = None, forceAbsolute: bool = True,
+                forceDirectory: bool = False) -> ButlerURI:
+        parsed: urllib.parse.ParseResult
+        dirLike: bool
+        subclass: Optional[Type] = None
+
+        # Record if we need to post process the URI components
+        # or if the instance is already fully configured
+        if isinstance(uri, str):
+            # Since local file names can have special characters in them
+            # we need to quote them for the parser but we can unquote
+            # later. Assume that all other URI schemes are quoted.
+            # Since sometimes people write file:/a/b and not file:///a/b
+            # we should not quote in the explicit case of file:
+            if "://" not in uri and not uri.startswith("file:"):
+                if ESCAPES_RE.search(uri):
+                    log.warning("Possible double encoding of %s", uri)
+                else:
+                    uri = urllib.parse.quote(uri)
+            parsed = urllib.parse.urlparse(uri)
+        elif isinstance(uri, urllib.parse.ParseResult):
+            parsed = copy.copy(uri)
+        elif isinstance(uri, ButlerURI):
+            parsed = copy.copy(uri._uri)
+            dirLike = uri.dirLike
+            # No further parsing required and we know the subclass
+            subclass = type(uri)
+        else:
+            raise ValueError(f"Supplied URI must be string, ButlerURI, or ParseResult but got '{uri!r}'")
+
+        if subclass is None:
+            # Work out the subclass from the URI scheme
+            if not parsed.scheme:
+                from .schemeless import ButlerSchemelessURI
+                subclass = ButlerSchemelessURI
+            elif parsed.scheme == "file":
+                from .file import ButlerFileURI
+                subclass = ButlerFileURI
+            elif parsed.scheme == "s3":
+                from .s3 import ButlerS3URI
+                subclass = ButlerS3URI
+            elif parsed.scheme.startswith("http"):
+                from .http import ButlerHttpURI
+                subclass = ButlerHttpURI
+            elif parsed.scheme == "resource":
+                # Rules for scheme names disasllow pkg_resource
```

Comment:
Minor typo in this comment. Maybe that was mine originally 😄

Reply:
Fixed!